plugins/net_plugin/net_plugin.cpp

```diff
@@ -3941,6 +3761,73 @@ namespace eosio {
       }
    }
 
+   digest_type gossip_bp_peers_message::bp_peer::digest() const {
+      digest_type::encoder enc;
+      fc::raw::pack(enc, my_impl->chain_id);
+      fc::raw::pack(enc, producer_name);
+      fc::raw::pack(enc, server_address);
+
+      return enc.result();
+   }
+
+   // called from connection strand
+   void connection::handle_message( gossip_bp_peers_message& msg ) {
+      if (!my_impl->bp_gossip_enabled())
+         return;
+
+      if (!my_impl->validate_gossip_bp_peers_message(msg)) {
+         peer_wlog( this, "bad gossip_bp_peers_message, closing");
+         no_retry = go_away_reason::fatal_other;
+         enqueue( go_away_message( fatal_other ) );
+         return;
+      }
+
+      // valid gossip peer connection
+      is_gossip_bp_connection = true;
+
+      assert(!msg.peers.empty()); // checked by validate_gossip_bp_peers_message()
+      if (msg.peers.size() == 1 && msg.peers[0].server_address.empty()) {
+         // initial message case, send back our entire collection
+         send_gossip_bp_peers_message();
+      } else {
+         bool diff = my_impl->update_gossip_bps(msg);
+         if (diff) { // update, let all our peers know about it
+            send_gossip_bp_peers_message_to_bp_peers();
```

Comment:
I think we should wait a few seconds before sending to all peers, so that we allow those peers to get the initial message from the source, preventing them from rebroadcasting the message again.

Reply:
Suppose:
- node A broadcasts a new peer to B, C, D, E.
- B receives it first and broadcasts to A, C, D, E
- C, D, E receive B's message (before A's) and each broadcast to 4 nodes.

If B had waited, C, D, and E would have received A's message first and would not have rebroadcasted (3rd step above)

Reply:
So a peer key update will cause a volley of (N-1)**2 gossip messages, if N is the number of finalizers. Could be an issue if N was to become significantly higher than 21.

Reply:
This scheme does not work for large number of finalizers. A new approach will be needed if/when the number of finalizers is expanded. Similarly vote aggregation will have to be handled differently. Presumably whatever scheme used for vote aggregation can also be used to connections.

Reply:
What do you think of that:

> One way to avoid this squaring of the messages would be if the gossip messages also included information to addresses which are notified, so the recipients would not send duplicate messages to these.

Reply:
You don't know peer actually received the message, only that it was sent. The processing of these messages seems extremely quick so not sure that is anything to worry about.